### PR TITLE
Theming + Carbonization

### PR DESCRIPTION
**Pull request by pinussilvestrus**
Closes https://github.com/bpmn-io/form-js/issues/515
Related to https://github.com/camunda/team-hto/issues/143

-----

Makes it possible to theme form viewer + editor by
* Mirror [Carbon Theme variables](https://carbondesignsystem.com/guidelines/themes/overview/)
* Make icons themeable
* Add simple starts for theming (`SINGLE_START=theme`)
* Add another package `@bpmn-io/form-js-carbon-styles` to make it easier to integrate custom form styles
  * Move carbon styles from Tasklist
  * Remove some styles as they are handled in the core styles now (mostly colors)

This is a step more towards Carbon in terms of theming, not implementing the exact component styles.

**Todo**
* [x] Gather feedback
* [x] Adjust changes from recent work stream (e.g. flexible forms)
* [x] Consider: Use carbonish icons in form-js number and select directly
* [x] Fixup wrong input border margins
* [x] Remove unnecessary styles (because they are handled by form-js already)
* [ ] Dry run to see whether something is broken: `npm run start:carbon`

**Theming**

image

image

**Carbon styles**

image

### CODE REVIEW

**Comment by pinussilvestrus** [comment]
We have a demo to review this one: Demo: https://demo-theming--camunda-form-playground.netlify.app/carbon (please focus on the canvas and ignore the other playground components for now). Feel free to have a look and report any styling issues you see!

File: packages/form-js-carbon-styles/src/carbon-styles.js
Comment (pinussilvestrus) [inline packages/form-js-carbon-styles/src/carbon-styles.js:5]: @vsgoulart I decided not to remove the icon overrides for numbers and selects and move it to the core as I still think it wouldn't look right to have such Carbon related look and feel in the base styles. I think it's okay to have the overrides as is.

File: packages/form-js-viewer/test/spec/Form.spec.js
Comment (vsgoulart) [inline packages/form-js-viewer/test/spec/Form.spec.js:134]: I think instead of checking for the class maybe we could add a visual regression test. What do you think?

File: packages/form-js-viewer/src/render/components/icons/Button.svg
Comment (vsgoulart) [inline packages/form-js-viewer/src/render/components/icons/Button.svg:1]: Just to confirm. Did you check if you could minimize these SVGs with that tool I mentioned once?

File: packages/form-js-carbon-styles/src/carbon-styles.js
Comment (vsgoulart) [reply packages/form-js-carbon-styles/src/carbon-styles.js:1]: To be honest I think we should have avoided it from the beginning :grimacing: 
I think as library maintainer we should be more mindful of bundle size and runtime cost impact (which `styled-components` has)
But for now let's go with it, we can always improve later as you said

File: packages/form-js-carbon-styles/src/carbon-styles.js
Comment (vsgoulart) [inline packages/form-js-carbon-styles/src/carbon-styles.js:888]: Instead of using interpolation you can just use CSS vars. I left because I was laxy
```suggestion
            margin: 0 var(--cds-spacing-05);
```

File: packages/form-js-carbon-styles/src/carbon-styles.js
Comment (vsgoulart) [inline packages/form-js-carbon-styles/src/carbon-styles.js:789]: These can also be replaced with CSS vars, just check for the CSS vars in the scope

File: packages/form-js-carbon-styles/src/carbon-styles.js
Comment (vsgoulart) [inline packages/form-js-carbon-styles/src/carbon-styles.js:754]: Very weird formatting :sweat_smile:

**Review by vsgoulart — COMMENTED** [review]
I'm getting some problems when linking it on Tasklist. Maybe we can check it together tomorrow afternoon
Besides that I just have some minors comments. I still will make a bit of a more thorough review on the env you created

**Review by vsgoulart — CHANGES_REQUESTED** [review]
1. The select styles are wrong. (I just noticed I probably forgot to update some of the styles like I promised :grimacing:) 
image
image

2. Disabled state of primary buttons (submit) are wrong. (I don't have these on Tasklist so you'll have copy them yourself)
image

3. Disabled taglists also look a bit weird
image
4. Disabled state of checklist are also wrong
image
5. Also number inputs
image
6. Date fields are also very messed up, but I guess it's because you included @Skaiir 's Flatpickr's styles?
image


I'll continue reviewing the styles tomorrow. I'll also try to take a look and update the stuff I update after you started working on this PR

File: packages/form-js-carbon-styles/test/spec/carbon-styles.spec.js
Comment (vsgoulart) [reply packages/form-js-carbon-styles/test/spec/carbon-styles.spec.js:47]: I think we can just remove these theme objects and just use CSS vars

File: packages/form-js-carbon-styles/src/carbon-styles.js
Comment (pinussilvestrus) [reply packages/form-js-carbon-styles/src/carbon-styles.js:1]: Let's directly add a comment to the styles so we follow up once we see it.

**Comment by pinussilvestrus** [comment]
> I'm getting some problems when linking it on Tasklist.

Maybe this refers to a similar problem I've got in the form-playground: https://github.com/camunda/form-playground/pull/56/files#diff-716e5d2ec328d34e2db57295d5e0c2432f8aa2611d42beedc8c56bcb61914564

There are some weird duplication problems with `createGlobalStyling` so it worked better when using the styles separately.

File: packages/form-js-viewer/src/render/components/icons/Button.svg
Comment (pinussilvestrus) [reply packages/form-js-viewer/src/render/components/icons/Button.svg:1]: Didn't check that for this PR. I think the time I added the changes, I didn't know about the tool 😅 Let's do that.

File: packages/form-js-viewer/test/spec/Form.spec.js
Comment (pinussilvestrus) [reply packages/form-js-viewer/test/spec/Form.spec.js:134]: I like that idea, good use case 👍

File: packages/form-js-carbon-styles/src/carbon-styles.js
Comment (pinussilvestrus) [reply packages/form-js-carbon-styles/src/carbon-styles.js:1]: Cf. https://github.com/bpmn-io/form-js/issues/633

**Comment by pinussilvestrus** [comment]
@vsgoulart did you see these styling issues in Tasklist? I don't see it in the playground demo.

File: packages/form-js-viewer/src/render/components/icons/Button.svg
Comment (pinussilvestrus) [reply packages/form-js-viewer/src/render/components/icons/Button.svg:1]: Just minified all of them.

File: packages/form-js-carbon-styles/src/carbon-styles.js
Comment (pinussilvestrus) [reply packages/form-js-carbon-styles/src/carbon-styles.js:791]: Do you mean we should remove all theme interpolations, so a ThemeProvider is not necessary anymore?

Comment (pinussilvestrus) [reply packages/form-js-carbon-styles/src/carbon-styles.js:789]: Ah yeah, cf. to this comment: https://github.com/bpmn-io/form-js/pull/557#discussion_r1179607872

Let's do that, relaxes the integration a lot 👍

Comment (pinussilvestrus) [reply packages/form-js-carbon-styles/src/carbon-styles.js:789]: By removing some `theme` occurrences, seems like I broke some styles: 
image

Maybe let's wait for the further review until I'm done with that.

Comment (pinussilvestrus) [reply packages/form-js-carbon-styles/src/carbon-styles.js:789]: Ok I fixed at least this issue. Feel free to continue your visual review 👍 I will aim to remove all the theme accesses, without breaking further things.

File: packages/form-js-carbon-styles/src/carbon-styles.js
Comment (pinussilvestrus) [reply packages/form-js-carbon-styles/src/carbon-styles.js]: Not sure which problem eslint has with this line 😅

File: packages/form-js-carbon-styles/src/carbon-styles.js
Comment (pinussilvestrus) [reply packages/form-js-carbon-styles/src/carbon-styles.js:795]: We just discussed that we can't remove the theme entirely as we need it for the background images (selects, numbers) - we can't use CSS variables over there.

File: packages/form-js-viewer/test/spec/Form.spec.js
Comment (pinussilvestrus) [reply packages/form-js-viewer/test/spec/Form.spec.js:134]: I added an issue for me: https://github.com/camunda/team-hto/issues/225

I think we can handle this separately.

File: packages/form-js-viewer/src/render/components/icons/Button.svg
Comment (vsgoulart) [reply packages/form-js-viewer/src/render/components/icons/Button.svg:1]: @pinussilvestrus Maybe we could document this somewhere?

**Review by vsgoulart — CHANGES_REQUESTED** [review]
Some other problems I found:
1. Margin on the searcheable select options is too big
image
2. Color of the value in a disabled number input is wrong
image
3. Disabled input numbers can be focused if you click on the control buttons
image

File: packages/form-js-carbon-styles/src/carbon-styles.js
Comment (pinussilvestrus) [inline packages/form-js-carbon-styles/src/carbon-styles.js:1085]: Imho, we should not include this here. Once we use these styles in the form-playground, this will break as you can place it from the palette, but won't see them on the canvas.

Couldn't we keep this particular one in Tasklist?

Comment (vsgoulart) [reply packages/form-js-carbon-styles/src/carbon-styles.js:1085]: Yep, I agree. It was just a copy&paste mistake.

**Pull request merged**
